File: examples/thymeleaf-java-example/src/main/java/de/tschuehly/example/thymeleafjava/web/action/ActionViewComponent.java
package de.tschuehly.example.thymeleafjava.web.action;

import de.tschuehly.example.thymeleafjava.core.ExampleService;
import de.tschuehly.spring.viewcomponent.core.action.*;
import de.tschuehly.spring.viewcomponent.core.component.ViewComponent;
import de.tschuehly.spring.viewcomponent.thymeleaf.ViewContext;
import jakarta.servlet.http.HttpServletRequest;
import org.springframework.web.bind.annotation.RequestParam;

import java.util.Map;

@ViewComponent
public class ActionViewComponent {
    private final ExampleService exampleService;

    public ActionViewComponent(ExampleService exampleService) {
        this.exampleService = exampleService;
    }

    public record ActionView(Integer counter, Map<Integer, String> itemList, Person person) implements ViewContext {
    }

    Integer counter = 0;

    public ViewContext render() {
        return new ActionView(counter, exampleService.itemList, person);
    }

    @GetViewAction(path = "/customPath/countUp")
    public ViewContext countUp() {
        counter += 1;
        return render();
    }


    record ActionFormDTO(
            String item
    ) {
    }

    @PostViewAction
    public ViewContext addItem(HttpServletRequest request,ActionFormDTO actionFormDTO) {
        exampleService.addItemToList(actionFormDTO.item);
        return render();
    }


    @DeleteViewAction
    public ViewContext deleteItem(Integer id) {
        exampleService.deleteItem(id);
        return render();
    }


    Person person = new Person(
            "Thomas", 23, "Ludwigsburg"
    );

    @PatchViewAction
    public ViewContext savePersonPatch(Person person) {
        this.person = person;
        return render();
    }

    @PutViewAction
    public ViewContext savePersonPut(HttpServletRequest request, Person person) {
        this.person = person; // TODO: ParameterBinding doesn't work
        return render();
    }

    public record Person(
            String name,
            Integer age,
            String location
    ) {
    }


}

File: thymeleaf/src/test/kotlin/de/tschuehly/spring/viewcomponent/thymeleaf/application/web/action/ActionViewComponent.kt
package de.tschuehly.spring.viewcomponent.thymeleaf.application.web.action

import de.tschuehly.spring.viewcomponent.core.*
import de.tschuehly.spring.viewcomponent.core.action.*
import de.tschuehly.spring.viewcomponent.core.component.ViewComponent
import de.tschuehly.spring.viewcomponent.thymeleaf.ViewContext
import de.tschuehly.spring.viewcomponent.thymeleaf.application.core.ExampleService

@ViewComponent
class ActionViewComponent(
    val exampleService: ExampleService
) {

    fun render() = ViewContext(
        "itemList" toProperty exampleService.someData,
        "counter" toProperty counter,
        "person" toProperty person
    )

    var counter = 0;

    @GetViewAction("/customPath/countUp")
    fun countUp(): ViewContext{
        counter += 1
        return render()
    }


    class ActionFormDTO(
        val item: String
    )
    @PostViewAction
    fun addItem(actionFormDTO: ActionFormDTO): ViewContext {
        exampleService.addItemToList(actionFormDTO.item)
        return render()
    }


    @DeleteViewAction
    fun deleteItem(id: Int): ViewContext{
        exampleService.deleteItem(id)
        return render()
    }


    var person = Person(
        name = "Thomas", age = 23, location = "Ludwigsburg"
    )

    @PatchViewAction
    fun savePersonPatch(person: Person): ViewContext {
        this.person = person
        return render()
    }
    @PutViewAction
    fun savePersonPut(person: Person): ViewContext {
        this.person = person
        return render()
    }
    class Person(
        val name: String,
        val age: Int,
        val location: String
    )
}